ledger/participant-integration-api/src/main/scala/platform/indexer/ha/PreemptableSequence.scala

```diff
@@ -130,23 +130,32 @@ object PreemptableSequence {
 
       override def retry[T](waitMillisBetweenRetries: Long, maxAmountOfRetries: Long)(
           body: => T
-      ): Future[T] =
+      )(retryGuard: Throwable => Boolean): Future[T] =
         go(body).transformWith {
           // since we check countdown to 0, starting from negative means unlimited retries
-          case Failure(ex) if maxAmountOfRetries == 0 =>
+          case Failure(ex) if retryGuard(ex) && maxAmountOfRetries == 0 =>
             logger.warn(
               s"Maximum amount of retries reached ($maxAmountOfRetries) failing permanently.",
               ex,
             )
             Future.failed(ex)
+
           case Success(t) => Future.successful(t)
-          case Failure(ex) =>
+
+          case Failure(ex) if retryGuard(ex) =>
             logger.debug(s"Retrying (retires left: ${if (maxAmountOfRetries < 0) "unlimited"
               else maxAmountOfRetries - 1}). Due to: ${ex.getMessage}")
             waitFor(waitMillisBetweenRetries).flatMap(_ =>
               // Note: this recursion is out of stack
-              retry(waitMillisBetweenRetries, maxAmountOfRetries - 1)(body)
+              retry(waitMillisBetweenRetries, maxAmountOfRetries - 1)(body)(retryGuard)
             )
+
+          case Failure(ex) =>
```

Comment:
Could have this condition first with `case Failure(ex) if !retryGuard(ex)` and save testing the guard twice (minor).

Reply:
I rewamped it a little bit so it is more balanced in terms of conditional branching